src/classes/Token/Token.ts

```diff
@@ -0,0 +1,109 @@
+// "why not a JWT"
+// There is literally no good reason for all of this
+// Why did I do this
+import * as crypto from "crypto";
+import Database from "../../db.js";
+
+export default class Token {
+    token: string;
+    type: "access"|"refresh";
+    expiresAt: Date;
+    createdAt: Date;
+
+    constructor(type: "access"|"refresh", expiry: Date = new Date(0)) {
+        let prefixPart = "JR";
+        let typePart;
+        switch (type) {
+            case "access":
+                typePart = "AC";
+                break;
+            case "refresh":
+                typePart = "RE";
+                break;
+            default:
+                throw new Error(`Unknown token type ${type}`)
+        }
+        let randomPart = crypto.randomBytes(48).toString("base64url");
+        let currentTimePart = Date.now().toString(36);
+        let expiryTimePart = expiry.getTime().toString(36);
+        this.token = [prefixPart, typePart, randomPart, currentTimePart, expiryTimePart].join(".");
+        this.expiresAt = expiry;
+        this.type = type;
+        this.createdAt = new Date(parseInt(currentTimePart, 36));
+    }
+
+    get expired() : boolean {
+        if (this.type === "refresh") return true;
+        return Date.now() > this.expiresAt.getTime()
+    }
+
+    /**
+     * Invalidate token-refresh token pair
+     * @returns {Promise<any>}
+     */
+    async invalidate() {
+        const database = new Database();
+        return await database.Token.deleteOne({$or: [{refresh: this.token}, {access: this.token}]});
+    }
+
+    async isActive() : Promise<any> {
+        const database = new Database();
+        let tokenDocument = await database.Token.findOne({$or: [{refresh: this.token}, {access: this.token}]}).populate("user");
+        if (!tokenDocument) return false;
+        if (this.type === "refresh") return tokenDocument;
+        return !this.expired ? tokenDocument : false;
+    }
+
+    /**
+     * Get a Token from an existing token
+     * @param token
+     * @returns {Token}
+     */
+    static from(token: string) {
+        let tokenInformation = Token.parse(token);
+        let oToken = new Token(tokenInformation.type, tokenInformation.expiresAt);
+        oToken.token = token;
+        oToken.createdAt = tokenInformation.createdAt;
+        return oToken;
+    }
+
+    toString() {
+        return this.token;
+    }
+
+    static parse(token: string|Token) {
+        // JR.RE.fz_z_MVQj8UQ3lY_UmrUjekE_2U4dn8d4WIkfHqJ7ZQRN-O1BNu6xfKINhKCrv1I.ljlioshm.ljlirzu1
+        // JR.AC.fz_z_MVQj8UQ3lY_UmrUjekE_2U4dn8d4WIkfHqJ7ZQRN-O1BNu6xfKINhKCrv1I.ljlioshm.ljlirzu1
+        // XX.YY.ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ.CCCCCCCC.EEEEEEEE
```

Comment:
The fuck happened here? Did you fall asleep?

Reply:
It explains how the tokens work

Reply:
it was a joke ;<